### ISSUE TITLE
Improve the calculation of similarity scores between answers and correct solutions

### ISSUE THREAD

**Issue by blmage** [blmage/duolingo-solution-viewer#2]
Currently, the similarity between answers and correct solutions is computed as-is, with only [Unicode normalization](https://unicode.org/faq/normalization.html) being applied. Therefore, accented letters and their unaccented counterparts are considered completely different characters.

While this is desirable when the user enters a "perfect" answer with regards to accents, it turns out that the results can get quite random in the contrary case.

A solution would be to compute two similarity scores, applying more or less normalization, then averaging them in a consistent way.

**Comment by blmage** [comment]
Rather leave the choice to the user of what is significant and what is not, as an option (see #25). This could include:
- case,
- accents,
- punctuation,
- spaces,
- word order (using the `diff` package - benchmark this on big lists of solutions to check whether this is a no-go).

At second glance, the latter should probably be a matching algorithm on its own, with the other options proposed for both (as of version `2.3.0`, only a few tweaks to the underlying functions would be required to make this work).

Only enable the Sørensen-Dice algorithm by default, and let the user specify a threshold over which each algorithm should not be used.

**Comment by tobiornottobi** [comment]
In my experience the order is completely off. There have been absurd sentences at the top (without any noticeable similarity) when the alphabetical sort gave me much more similar answers.

**Comment by blmage** [comment]
@tobiornottobi Could you please send one or two screenshots with examples of such behavior?

I'm only aware of this happening with missing or different diacritics, but I'll increase the priority of this issue if this happens to be more widespread.

Thanks!

**Comment by tobiornottobi** [comment]
@blmage Yes, I can. One thing I have to add: I wasn't sure if .* sort↓ button toggles the other option or says which option is currently active. The results weren't sorted alphabetically, so maybe it's actually the alphabetical sort that is broken for me.
I haven't gotten absurd suggestions this time – because the accepted answers are all reasonable and similar, but I still don't understand the order.
This is neither sorted by similarity nor alphabetically. Unless only the first word is taken into account.
![image](https://user-images.githubusercontent.com/44438257/92306420-3c584100-ef8f-11ea-9456-d4965622e67c.png)
This makes sense similarity-wise:
![image](https://user-images.githubusercontent.com/44438257/92306399-25b1ea00-ef8f-11ea-81cb-cb38d9bc9f9f.png)

I'll try to remember making a screenshot in the future.

**Comment by blmage** [comment]
@tobiornottobi Thanks for the screenshots!

The UI reflects the current state, so when "**Alphabetical sort ↓**" is displayed, solutions are/should be sorted alphabetically and in descending order.

The order on the first screenshot seems correct, apart from the two solutions at the top, but I couldn't reproduce the same result in isolation (when testing the comparison algorithm, "ä" comes before "b", as expected).

Could you point me to a skill in the Norwegian tree that uses a lot of accented words? (I'll try to reproduce it from there instead)

**Comment by tobiornottobi** [comment]
@blmage Thank you. :)
The screenshot was from the Swedish tree. I can't search at the moment unfortunately.

**Comment by blmage** [comment]
My bad! In the case of Swedish then, this seems to be the expected behavior:

> In addition to the basic twenty-six letters, A–Z, the Swedish alphabet includes Å, Ä, and Ö at the end. They are distinct letters in Swedish, and are sorted after Z as shown above.

##### [Wikipedia](https://en.wikipedia.org/wiki/Swedish_alphabet#Accented_letters)
